app/models/ability.rb

```diff
@@ -0,0 +1,71 @@
+class Ability
+  include CanCan::Ability
+
+  def initialize(user)
+    user ||= User.new
+
+    can :manage, :all if user.is_admin?
+    can [:manage_item, :manage_content, :manage_desc_metadata, :register_apo], :all if user.is_manager?
+    can [:view_item, :view_metadata, :view_content], :all if user.is_viewer?
```

Comment:
The apo manager roles limits the user to only those items they have access to by virtue of item governance (the purpose of roles), while sdr-admin grants all possible permissions for all objects (the sdr-* roles are all meant to be repository-wide, though we only have sdr-admin to date). I can’t speak to the actual implementation, but at a practical level recognizing super powers gained considerable efficiency by bypassing the user-to-roles mapping via workgroups for all APOs (hundreds and hundreds of them) and the cost of filtering all facets and searches.  I don’t know if there are similar efficiencies at the granular level of checking permissions for individual actions/button/etc, but I imagine there could be in some cases.

Does this help, or are there other fuzzy aspects?


> On Feb 26, 2016, at 2:27 PM, Chris Beer <notifications@github.com> wrote:
> 
> In app/models/ability.rb:
> 
> > @@ -0,0 +1,71 @@
> > +class Ability
> > +  include CanCan::Ability
> > +
> > +  def initialize(user)
> > +    user ||= User.new
> > +
> > +    can :manage, :all if user.is_admin?
> > +    can [:manage_item, :manage_content, :manage_desc_metadata, :register_apo], :all if user.is_manager?
> > +    can [:view_item, :view_metadata, :view_content], :all if user.is_viewer?
> 
> Yes, permissions cascade as expected, and short-circuit when it gets an affirmative response.
> 
> Agreed, the difference between the application admin and manager roles is very unclear. Maybe @LynnMcRae can shed some light.
> 
> —
> Reply to this email directly or view it on GitHub.
>